hackDetail.js

```diff
@@ -5,6 +5,9 @@ const url = `mongodb://${process.env.MONGO_USER}:${process.env.MONGO_PASSWORD}@d
 
 export const detail = async (event, context) => {
   const id = event.pathParameters.id;
+  const userId = event.queryStringParameters.userId;
```

Comment:
can we identify this from the token we with the request?

Reply:
I have tried to but because the JWT is from cogito, I can't modify the token to include our mongodb userId without invalidating the token, also can't change that in front end as that will also invalidate token=(

Reply:
initially, after the user is logged in we return an object that contains jwt token along with userId fetched from mongodb

Reply:
Is there a Cognito id in the token? Do we store that against the user we store (we should be... and we should index that field also to make it fast to lookup by that id)? We could use instead of the mongodb id in all places we reference a user, but populating the users may need to change... https://mongoosejs.com/docs/populate.html#query-conditions)

Reply:
now that I think about it, we could extract info from the jwt token and make our own! WIll look into that.

Reply:
You can also set your own _id field in mongo, does it have to be a mongodb id type, thats just the default if you dont set an _id field in your document.

Reply:
I will try to implement our custom token for now, not sure if aws will throw a fit if it detected that the userId in the jwt is not its own.

Reply:
I don't think you should go down the path of creating your own token, for it to work you would need to sign your jwt with a private key to be able to trust it hasn't been tampered with (which is what you're actually doing by trying to add information to it.

Reply:
Unfortunately in the accessToken issued by Cognito, it doesn't have the userId from the Cognito user pool. Creating our jwt doesn't look viable either as our middleware use AWS signature verification.

At the moment there is a pre-hook in aws to modify token however it can't modify access token only id token. I feel at this moment the query string is adequate for this particular GET request

Reply:
thats unfortunate. I feel like we/I'm missing something with cognito user pools and iam

https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-accessing-resources-api-gateway-and-lambda.html

They make it sound and look so simple. Perhaps we are just using it wrong?

I wonder if things might be easier with an AWS managed database rather than our mongoDB one.

Reply:
From one of the example full-stack app from serverless-stack.com. It does seem pretty simple with the conjunction of Amplify client-side code along with AWS identify pool service to access DynamoDB.

You will only need to declare this in serverless.yml

```yaml
iamRoleStatements:
  - Effect: Allow
    Action:
      - dynamodb:DescribeTable
      - dynamodb:Query
      - dynamodb:Scan
      - dynamodb:GetItem
      - dynamodb:PutItem
      - dynamodb:UpdateItem
      - dynamodb:DeleteItem
    Resource: "arn:aws:dynamodb:ap-southeast-2:019137212646:table/notes"
```

and then put the authorizer function like this:

```yaml
functions:
  billing:
    # Defines an HTTP API endpoint that calls the main function in billing.js
    # - path: url path is /billing
    # - method: POST request
    handler: ./services/notes/billing.main
    events:
      - http:
          path: billing
          method: post
          cors: true
          authorizer: aws_iam
```

In future projects perhaps with serverless backend, we can integrate front end with Amplify and that will seamlessly work well with Cognito user pool service along with Identity and services like DynamoDB and S3.